rust/otap-dataflow/crates/otap/src/retry_processor.rs

```diff
@@ -323,208 +234,212 @@ pub fn create_retry_processor(
         let metrics = pipeline_ctx.register_metrics::<RetryProcessorMetrics>();
-        Self {
-            config,
-            pending_messages: HashMap::new(),
-            next_message_id: 1,
-            last_cleanup_time: Instant::now(),
-            metrics: Some(metrics),
-        }
-    }
-
-    async fn acknowledge(&mut self, ack: AckMsg<OtapPdata>) -> Result<(), Error> {
-        let rstate: RetryState = ack.calldata.try_into()?;
-        let id = rstate.id;
-        if let Some(_removed) = self.pending_messages.remove(&id) {
-            log::debug!("Acknowledged and removed message with ID: {id}");
-            if let Some(m) = self.metrics.as_mut() {
-                m.msgs_acked.inc();
-            }
-        } else {
-            log::warn!("Attempted to acknowledge non-existent message with ID: {id}");
-        }
-        Ok(())
+        Self { config, metrics }
     }
 
-    async fn handle_nack(
+    async fn handle_ack(
         &mut self,
-        nack: NackMsg<OtapPdata>,
-        _effect_handler: &mut EffectHandler<OtapPdata>,
+        ack: AckMsg<OtapPdata>,
+        effect_handler: &mut EffectHandler<OtapPdata>,
     ) -> Result<(), Error> {
-        let rstate: RetryState = nack.calldata.try_into()?;
+        let signal = ack.accepted.signal_type();
+        let calldata = ack.calldata.clone();
 
-        let id = rstate.id;
-        if let Some(m) = self.metrics.as_mut() {
-            m.nacks_received.inc();
-        }
-        if let Some(mut pending) = self.pending_messages.remove(&id) {
-            pending.retry_count += 1;
-            pending.last_error = nack.reason;
-
-            if pending.retry_count <= self.config.max_retries {
-                let delay_ms = (self.config.initial_retry_delay_ms as f64
-                    * self
-                        .config
-                        .backoff_multiplier
-                        .powi(pending.retry_count as i32 - 1))
-                .min(self.config.max_retry_delay_ms as f64) as u64;
-
-                pending.next_retry_time = Instant::now() + Duration::from_millis(delay_ms);
-                let retry_count = pending.retry_count;
-                let _previous = self.pending_messages.insert(id, pending);
-                if let Some(m) = self.metrics.as_mut() {
-                    m.retry_attempts.inc();
-                }
-                log::debug!("Scheduled message {id} for retry attempt {retry_count}");
-            } else {
-                log::error!(
-                    "Message {} exceeded max retries ({}), dropping. Last error: {}",
-                    id,
-                    self.config.max_retries,
-                    pending.last_error
-                );
-                if let Some(m) = self.metrics.as_mut() {
-                    m.msgs_dropped_exceeded_retries.inc();
-                }
+        effect_handler.notify_ack(ack).await?;
+
+        // Forward the ACK upstream to complete the request chain
```

Comment:
The comment should be above the `notify_ack` call.

```suggestion
        // Forward the ACK upstream to complete the request chain
        effect_handler.notify_ack(ack).await?;

```